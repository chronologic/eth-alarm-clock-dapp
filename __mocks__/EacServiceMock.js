export const MOCKED_TRANSACTIONS = [
  '0x123306090abab3a6e1400e9345bc60c78a8bef57',
  '0x456306090abab3a6e1400e9345bc60c78a8bef57'
];

export const eacService = {
  calcEndowment(num) {
    return num;
  },
  getActiveContracts: () => {
    return {};
  },
  requestFactory: () =>
    Promise.resolve({
      getRequestCreatedEvents() {
        return MOCKED_TRANSACTIONS.map(tx => ({
          args: {
            request: tx,
            params: []
          }
        }));
      }
    }),
  scheduler: () => Promise.resolve({}),

  transactionRequest(address) {
    return {
      address
    };
  },
  getTransactionsEventsForAddresses() {
    return {};
  },
  Util: {
    getBlockNumber() {
      return 1;
    },
    getTimestampForBlock: () => 123
  },
  RequestData() {
    return {};
  },
  getTotalEthTransferred: () => {
    return 1000;
  }
};
