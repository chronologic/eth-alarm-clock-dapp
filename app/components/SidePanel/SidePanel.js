import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { NavLink } from 'react-router-dom';
import { observer, inject } from 'mobx-react';
import { isRunningInElectron } from '../../lib/electron-util';

@inject('web3Service')
@inject('keenStore')
@observer
class SidePanel extends Component {
  constructor() {
    super();
    this.state = {
      isElectron: isRunningInElectron()
    };
  }

  componentDidMount() {
    const { jQuery } = window;

    if (jQuery && jQuery.Pages) {
      jQuery.Pages.init();
    }
  }

  isUrlActive(url, type = 'thumbnail', substring = false) {
    const currentUrl = this.props.location.pathname;
    const cls = type === 'thumbnail' ? 'active' : 'text-white';

    if (substring) {
      return currentUrl.includes(url) ? cls : '';
    } else {
      return currentUrl === url ? cls : '';
    }
  }

  render() {
    const titleClasses = 'title ';
    const subtitleClasses = 'sub-title ';
    const thumbnailClasses = 'icon-thumbnail ';

    const entryList = [
      {
        title: 'Schedule',
        titleClasses: titleClasses + this.isUrlActive('/', 'title'),
        thumbnailClasses: thumbnailClasses + this.isUrlActive('/', 'thumbnail')
      },
      {
        title: 'Transactions',
        titleClasses:
          titleClasses + subtitleClasses + this.isUrlActive('/transactions', 'title', true),
        thumbnailClasses: thumbnailClasses + this.isUrlActive('/transactions', 'thumbnail', true)
      },
      {
        title: 'TimeNode',
        titleClasses: titleClasses + this.isUrlActive('/timenode', 'title'),
        thumbnailClasses: thumbnailClasses + this.isUrlActive('/timenode', 'thumbnail')
      },
      {
        title: 'Faucet',
        titleClasses: titleClasses + this.isUrlActive('/faucet', 'title'),
        thumbnailClasses: thumbnailClasses + this.isUrlActive('/faucet', 'thumbnail')
      }
    ];

    const { isElectron } = this.state;

    return (
      <nav className="page-sidebar" data-pages="sidebar">
        <div className="sidebar-header">
          <img
            src="img/logo-white.png"
            alt="logo"
            className="brand"
            data-src="img/logo-white.png"
            height="36"
          />
          <div className="sidebar-header-controls">
            <button
              type="button"
              className="btn btn-link d-lg-inline-block d-xlg-inline-block d-md-inline-block d-sm-none d-none"
              data-toggle-pin="sidebar"
            >
              <i className="fa fa-lock" />
            </button>
          </div>
        </div>
        <div className="sidebar-menu">
          <ul className="menu-items">
            {!isElectron && (
              <li className="m-t-30 ">
                <NavLink to="/">
                  <span className={entryList[0].titleClasses}>{entryList[0].title}</span>
                  <span className={entryList[0].thumbnailClasses}>
                    <i className="pg-calender" />
                  </span>
                </NavLink>
              </li>
            )}
            {!isElectron && (
              <li>
                <a href="#" onClick={e => e.preventDefault()}>
                  <span className={entryList[1].titleClasses}>{entryList[1].title}</span>
                  <span className={entryList[1].thumbnailClasses}>
                    <i className="pg-charts" />
                  </span>
                </a>

                <ul className="sub-menu">
                  <li>
                    <NavLink to="/transactions/scheduled">
                      <span className="title">Scheduled</span>
                      <span className="icon-thumbnail">
                        <i className="pg-plus_circle" />
                      </span>
                    </NavLink>
                  </li>
                  <li>
                    <NavLink to="/transactions/completed">
                      <span className="title">Completed</span>
                      <span className="icon-thumbnail">
                        <i className="fa fa-check" />
                      </span>
                    </NavLink>
                  </li>
                </ul>
              </li>
            )}
            <li>
              <NavLink to={isElectron ? '/timenode?mode=electron' : '/timenode'}>
                <span className={entryList[2].titleClasses}>{entryList[2].title}</span>
                <span className={entryList[2].thumbnailClasses}>
                  <i className="fa fa-sitemap" />
                </span>
              </NavLink>
            </li>
            {!isElectron && (
              <li>
                <a
                  href="https://alpha.chronologic.network/debt/"
                  target="_blank"
                  rel="noopener noreferrer"
                >
                  <span className="title">Debt Smart Contract</span>
                  <span className="icon-thumbnail">
                    <i className="fab fa-ethereum" />
                  </span>
                </a>
              </li>
            )}
            {!isElectron && (
              <li>
                <a
                  href="https://alpha.chronologic.network/chronos/"
                  target="_blank"
                  rel="noopener noreferrer"
                >
                  <span className="title">Day Token Contract</span>
                  <span className="icon-thumbnail">
                    <i className="far fa-clock" />
                  </span>
                </a>
              </li>
            )}
            {!isElectron && (
              <li>
                <NavLink to="/faucet">
                  <span className={entryList[3].titleClasses}>{entryList[3].title}</span>
                  <span className={entryList[3].thumbnailClasses}>
                    <i className="fas fa-tint" />
                  </span>
                </NavLink>
              </li>
            )}
            <li>
              <a
                href="https://blog.chronologic.network/chronos-platform/home"
                target="_blank"
                rel="noopener noreferrer"
              >
                <span className="title">Help</span>
                <span className="icon-thumbnail">
                  <i className="far fa-question-circle" />
                </span>
              </a>
            </li>

            <hr id="sidebar-separator" className="d-md-block d-lg-none mx-4" />

            <li className="d-md-block d-lg-none">
              <div className="container py-2">
                <div className="row p-l-20 p-r-15">
                  <div className="col-8 px-0">
                    <span className="active-timenodes">Active TimeNodes</span>
                  </div>
                  <div className="col-4 px-0 text-right">
                    <span className="timenode-count col-6">
                      {this.props.keenStore.activeTimeNodes}
                    </span>
                  </div>
                </div>
              </div>
            </li>

            <li className="d-md-block d-lg-none">
              <div className="container py-2">
                <div className="row p-l-20 p-r-15">
                  <div className="col-8 px-0">
                    <span className="active-timenodes">Current Block</span>
                  </div>
                  <div className="col-4 px-0 text-right">
                    <span className="timenode-count col-6">
                      {this.props.web3Service.latestBlockNumber}
                    </span>
                  </div>
                </div>
              </div>
            </li>
          </ul>
          <div className="clearfix" />
        </div>
      </nav>
    );
  }
}

SidePanel.propTypes = {
  web3Service: PropTypes.any,
  keenStore: PropTypes.any,
  location: PropTypes.object.isRequired
};

export default SidePanel;
