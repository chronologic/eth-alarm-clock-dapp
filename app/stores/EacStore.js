import { autorun, observable } from 'mobx';
import EacCounter from 'eac-counter';

export default class EacStore {
  @observable
  contracts = {};

  @observable
  totalEthTransferred = null;
  @observable
  totalUsdTransferred = null;

  constructor(eacService, featuresService, web3Service) {
    this._eac = eacService;
    this._featuresService = featuresService;
    this._web3Service = web3Service;

    autorun(() => this.refreshCurrentState());
  }

  async refreshCurrentState() {
    if (!this._featuresService._isCurrentNetworkSupported) {
      return;
    }

    await this._web3Service.init();

    this.contracts = await this._eac.getActiveContracts();

    const eacCounter = new EacCounter();

    await eacCounter.enableUSDFetching(process.env.NOMICS_API_KEY);

    const { eth, usd } = await eacCounter.getTotalTransferred();

    this.totalEthTransferred = Math.round(eth);
    this.totalUsdTransferred = Math.round(usd);
  }

  getFormattedUSDTranferred() {
    const currencyFormatter = new Intl.NumberFormat('en-US', {
      style: 'currency',
      currency: 'USD',
      minimumFractionDigits: 0
    });
    return currencyFormatter.format(this.totalUsdTransferred);
  }
}
