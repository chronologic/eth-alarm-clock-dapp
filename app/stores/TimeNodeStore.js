import { observable, computed } from 'mobx';
import CryptoJS from 'crypto-js';
import ethereumJsWallet from 'ethereumjs-wallet';
import EacWorker from 'worker-loader!../js/eac-worker.js';
import { EAC_WORKER_MESSAGE_TYPES } from '../js/eac-worker-message-types';
import { showNotification } from '../services/notification';
import { LOGGER_MSG_TYPES, LOG_TYPE } from '../lib/worker-logger.js';
import { isMyCryptoSigValid, isSignatureValid, parseSig, SIGNATURE_ERRORS } from '../lib/signature';
import { getDAYBalance } from '../lib/timenode-util';
import { Config } from '@ethereum-alarm-clock/timenode-core';
import { isRunningInElectron } from '../lib/electron-util';
import { Networks } from '../config/web3Config';

/*
 * TimeNode classification based on the number
 * of DAY tokens held by the owner.
 */
export class TIMENODE_STATUS {
  static MASTER_CHRONONODE = {
    name: 'Master ChronoNode',
    minBalance: 3333
  };
  static CHRONONODE = {
    name: 'ChronoNode',
    minBalance: 888
  };
  static TIMENODE = {
    name: 'TimeNode',
    minBalance: 333
  };
  static DISABLED = 'Disabled';
  static LOADING = 'Loading';
}

// 2 minute as milliseconds
const STATUS_UPDATE_INTERVAL = 2 * 60 * 1000;
const LOG_CAP = 1000;

export default class TimeNodeStore {
  @observable
  walletKeystore = '';
  @observable
  attachedDAYAccount = '';
  @observable
  scanningStarted = false;
  @observable
  claiming = false;

  @observable
  unlocked = false;

  @observable
  basicLogs = [];
  @observable
  detailedLogs = [];

  @observable
  logType = LOG_TYPE.BASIC;
  @computed
  get logs() {
    return this.logType === LOG_TYPE.BASIC ? this.basicLogs : this.detailedLogs;
  }

  @observable
  successfulExecutions = null;
  @observable
  failedExecutions = null;
  @observable
  successfulClaims = null;
  @observable
  failedClaims = null;
  @observable
  discovered = null;

  @observable
  balanceETH = null;
  @observable
  balanceDAY = null;
  @observable
  isTimeMint = null;

  @computed
  get nodeStatus() {
    const { MASTER_CHRONONODE, CHRONONODE, TIMENODE, DISABLED, LOADING } = TIMENODE_STATUS;

    if (this.balanceDAY === null) {
      return LOADING;
    }

    if (this.balanceDAY >= MASTER_CHRONONODE.minBalance) {
      return MASTER_CHRONONODE.name;
    } else if (this.balanceDAY >= CHRONONODE.minBalance) {
      return CHRONONODE.name;
    } else if (this.balanceDAY >= TIMENODE.minBalance || this.isTimeMint) {
      return TIMENODE.name;
    } else {
      return DISABLED;
    }
  }

  @observable
  bounties = null;
  @observable
  costs = null;
  @observable
  profit = null;

  @computed
  get economicStrategy() {
    const load = strategy => {
      const loaded = this._storageService.load(strategy);
      return loaded ? loaded : Config.DEFAULT_ECONOMIC_STRATEGY[strategy].toString();
    };

    return {
      maxDeposit: load('maxDeposit'),
      minBalance: load('minBalance'),
      minProfitability: load('minProfitability'),
      maxGasSubsidy: load('maxGasSubsidy')
    };
  }

  // If a TimeNode has selected a custom provider URL
  // it will be stored in this variable
  @observable
  customProviderUrl = null;
  @observable
  providerBlockNumber = null;

  netId = null;

  get network() {
    const customNetId = this.getCustomProvider().id;
    const currentNetId = customNetId ? customNetId : this._web3Service.network.id;
    if (!Networks[currentNetId]) {
      return this.getCustomProvider();
    }
    return Networks[currentNetId];
  }

  eacWorker = null;

  _keenStore = null;
  _storageService = null;
  _timeNodeStatusCheckIntervalRef = null;

  updateStatsInterval = null;
  updateBalancesInterval = null;
  getNetworkInfoInterval = null;

  constructor(eacService, web3Service, keenStore, storageService) {
    this._eacService = eacService;
    this._web3Service = web3Service;
    this._keenStore = keenStore;
    this._storageService = storageService;

    if (this._storageService.load('attachedDAYAccount') !== null)
      this.attachedDAYAccount = this._storageService.load('attachedDAYAccount');
    if (this._storageService.load('tn') !== null)
      this.walletKeystore = this._storageService.load('tn');
    if (this._storageService.load('claiming')) this.claiming = true;

    this.updateStats = this.updateStats.bind(this);
    this.updateBalances = this.updateBalances.bind(this);
    this.getNetworkInfo = this.getNetworkInfo.bind(this);
  }

  async unlockTimeNode(password) {
    if (this.walletKeystore && password) {
      this.unlocked = true;
      await this.startClient(this.walletKeystore, password);
      if (localStorage.getItem('isTimenodeScanning')) {
        await this.startScanning();
      }
    } else {
      this.unlocked = false;
      showNotification('Unable to unlock the TimeNode. Please try again');
    }
  }

  getWorkerOptions(keystore, keystorePassword) {
    return {
      network: this.network,
      customProviderUrl: this.customProviderUrl,
      keystore: [this.decrypt(keystore)],
      keystorePassword,
      dayAccountAddress: this.getAttachedDAYAddress(),
      logfile: 'console',
      logLevel: 1,
      milliseconds: 15000,
      autostart: false,
      scan: 950, // ~65min on kovan
      repl: false,
      browserDB: true,
      economicStrategy: this.economicStrategy,
      claiming: this.claiming
    };
  }

  stopIntervals() {
    clearInterval(this.updateStatsInterval);
    clearInterval(this.updateBalancesInterval);
    clearInterval(this.getNetworkInfoInterval);
  }

  startIntervals() {
    this.updateStats();
    this.updateStatsInterval = setInterval(this.updateStats, 5000);

    this.updateBalances();
    this.updateBalancesInterval = setInterval(this.updateBalances, 15000);

    this.getNetworkInfo();
    this.getNetworkInfoInterval = setInterval(this.getNetworkInfo, 15000);
  }

  startWorker(options) {
    return new Promise(resolve => {
      this.eacWorker = new EacWorker();

      this.eacWorker.onmessage = async event => {
        const { type, value } = event.data;
        const getValuesIfInMessage = values => {
          values.forEach(value => {
            if (event.data[value] !== null) {
              this[value] = event.data[value];
            }
          });
        };

        switch (type) {
          case EAC_WORKER_MESSAGE_TYPES.STARTED:
            this.stopIntervals();
            this.startIntervals();

            resolve();
            break;

          case EAC_WORKER_MESSAGE_TYPES.LOG:
            this.handleLogMessage(value);
            break;

          case EAC_WORKER_MESSAGE_TYPES.UPDATE_STATS:
            getValuesIfInMessage([
              'bounties',
              'costs',
              'profit',
              'successfulClaims',
              'failedClaims',
              'successfulExecutions',
              'failedExecutions',
              'discovered'
            ]);
            break;

          case EAC_WORKER_MESSAGE_TYPES.UPDATE_BALANCES:
            getValuesIfInMessage(['balanceETH', 'balanceDAY', 'isTimeMint']);
            break;

          case EAC_WORKER_MESSAGE_TYPES.CLEAR_STATS:
            showNotification('Cleared the stats.', 'success');
            this.updateStats();
            break;

          case EAC_WORKER_MESSAGE_TYPES.GET_NETWORK_INFO:
            getValuesIfInMessage(['providerBlockNumber', 'netId']);
            if (this._keenStore.timeNodeSpecificProviderNetId != this.netId) {
              this._keenStore.setTimeNodeSpecificProviderNetId(this.netId);
              await this._keenStore.refreshActiveTimeNodesCount();
            }
            break;

          case EAC_WORKER_MESSAGE_TYPES.RECEIVED_CLAIMED_NOT_EXECUTED_TRANSACTIONS:
            this._getClaimedNotExecutedTransactionsPromiseResolver(event.data['transactions']);
            break;
        }
      };

      this.eacWorker.postMessage({
        type: EAC_WORKER_MESSAGE_TYPES.START,
        options
      });
    });
  }

  async getClaimedNotExecutedTransactions() {
    this.eacWorker.postMessage({
      type: EAC_WORKER_MESSAGE_TYPES.GET_CLAIMED_NOT_EXECUTED_TRANSACTIONS
    });

    return new Promise(resolve => {
      this._getClaimedNotExecutedTransactionsPromiseResolver = resolve;
    });
  }

  pushToLog(logs, log) {
    if (logs.length === LOG_CAP) logs.shift();
    logs.push(log);
  }

  handleLogMessage(log) {
    if (log.type === LOGGER_MSG_TYPES.CACHE) return;
    if (log.type !== LOGGER_MSG_TYPES.DEBUG) {
      this.pushToLog(this.basicLogs, log);
    }

    this.pushToLog(this.detailedLogs, log);
  }

  sendActiveTimeNodeEvent() {
    if (this.scanningStarted) {
      this._keenStore.sendActiveTimeNodeEvent(this.getMyAddress(), this.getAttachedDAYAddress());
    }
  }

  async startScanning() {
    if (this.nodeStatus === TIMENODE_STATUS.DISABLED) {
      return;
    }

    this.scanningStarted = true;

    this.sendActiveTimeNodeEvent();

    this._timeNodeStatusCheckIntervalRef = setInterval(
      () => this.sendActiveTimeNodeEvent(),
      STATUS_UPDATE_INTERVAL
    );

    this.eacWorker.postMessage({
      type: EAC_WORKER_MESSAGE_TYPES.START_SCANNING
    });

    this.updateStats();
    this._storageService.save('isTimenodeScanning', true);
  }

  stopScanning() {
    this.scanningStarted = false;

    if (this._timeNodeStatusCheckIntervalRef) {
      clearInterval(this._timeNodeStatusCheckIntervalRef);
    }

    if (this.eacWorker) {
      this.eacWorker.postMessage({
        type: EAC_WORKER_MESSAGE_TYPES.STOP_SCANNING
      });
    }
    this._storageService.remove('isTimenodeScanning');
  }

  encrypt(message) {
    return CryptoJS.AES.encrypt(message, '88e19245648ba7616099fbd6595d120d');
  }

  decrypt(message) {
    if (typeof message !== 'string') {
      message = message.toString();
    }
    const bytes = CryptoJS.AES.decrypt(message, '88e19245648ba7616099fbd6595d120d');
    return bytes.toString(CryptoJS.enc.Utf8);
  }

  /*
   * Starts the timenode client.
   * Immediately starts executing transactions and outputting logs.
   */
  async startClient(keystore, password) {
    await this._web3Service.init();

    await this.startWorker(this.getWorkerOptions(keystore, password));
  }

  setKeyStore(keystore) {
    this.walletKeystore = keystore;
    this._storageService.save('tn', keystore);
  }

  setCustomProvider(id, endpoint) {
    this.customProviderUrl = endpoint;
    this._storageService.save('selectedProviderId', id);
    this._storageService.save('selectedProviderEndpoint', endpoint);

    this.stopScanning();

    // Reload the page so that the changes are refreshed
    if (isRunningInElectron()) {
      // Workaround for getting the Electron app to reload
      // since the regular reload results in a blank screen
      window.location.href = '/index.html';
    } else {
      window.location.reload();
    }
  }

  getCustomProvider() {
    return {
      id: parseInt(this._storageService.load('selectedProviderId')),
      endpoint: this._storageService.load('selectedProviderEndpoint')
    };
  }

  getMyAddress() {
    if (this.walletKeystore) {
      const ks = this.decrypt(this.walletKeystore);
      const address = JSON.parse(ks).address;

      if (address && address.indexOf('0x') === -1) {
        return '0x' + address;
      }

      return address;
    }

    return '';
  }

  getAttachedDAYAddress() {
    const encryptedAddress = this._storageService.load('attachedDAYAccount');
    if (encryptedAddress) {
      return this.decrypt(encryptedAddress);
    } else {
      return '';
    }
  }

  sendMessageWorker(messageType) {
    if (this.eacWorker) {
      this.eacWorker.postMessage({
        type: messageType
      });
    }
  }

  getNetworkInfo() {
    this.sendMessageWorker(EAC_WORKER_MESSAGE_TYPES.GET_NETWORK_INFO);
  }

  updateStats() {
    this.sendMessageWorker(EAC_WORKER_MESSAGE_TYPES.UPDATE_STATS);
  }

  updateBalances() {
    this.sendMessageWorker(EAC_WORKER_MESSAGE_TYPES.UPDATE_BALANCES);
  }

  clearStats() {
    this.sendMessageWorker(EAC_WORKER_MESSAGE_TYPES.CLEAR_STATS);
    this.basicLogs = [];
    this.detailedLogs = [];
  }

  /*
   * Attaches a DAY-token-holding account to the session
   * as a proof-of-ownership of DAY tokens.
   * If it contains DAY tokens, it allows the usage of TimeNodes.
   */
  async attachDayAccount(sigObject) {
    try {
      const signature = parseSig(sigObject);

      // First check using default sig check - if doesn't work use MyCrypto's
      const validSig = isSignatureValid(signature) ? true : isMyCryptoSigValid(signature);

      if (!validSig) throw SIGNATURE_ERRORS.INVALID_SIG;

      const { balanceDAY, mintingPower } = await getDAYBalance(
        this.network,
        this._web3Service.getWeb3FromProviderUrl(this.network.endpoint),
        signature.address
      );

      this.balanceDAY = balanceDAY.toNumber();
      this.isTimeMint = mintingPower > 0;

      const encryptedAttachedAddress = this.encrypt(signature.address);

      if (this.nodeStatus !== TIMENODE_STATUS.DISABLED) {
        this._storageService.save('attachedDAYAccount', encryptedAttachedAddress);
        this.attachedDAYAccount = encryptedAttachedAddress;
        showNotification('Success.', 'success');
      } else {
        showNotification('Not enough DAY tokens. Current balance: ' + balanceDAY.toString());
      }
    } catch (error) {
      if (error == `TypeError: Cannot read property 'dayTokenAddress' of undefined`) {
        showNotification('Unsupported custom provider.');
      } else {
        showNotification(error);
      }
    }
  }

  saveClaimingStrategy(economicStrategy) {
    if (this.claiming) {
      this._storageService.save('claiming', true);
    } else {
      this._storageService.remove('claiming');
    }

    const numberFromString = string => this._web3Service.web3.toWei(string, 'ether');
    for (let key of Object.keys(economicStrategy)) {
      if (economicStrategy[key]) {
        const value =
          key === 'maxGasSubsidy' ? economicStrategy[key] : numberFromString(economicStrategy[key]);
        this._storageService.save(key, value);
      } else {
        this._storageService.remove(key);
      }
    }
  }

  hasStorageItems(itemList) {
    for (let item of itemList) {
      if (!this._storageService.load(item)) {
        return false;
      }
    }
    return true;
  }

  async restart(password) {
    this.stopScanning();
    this.stopIntervals();
    this.eacWorker = null;
    await this.startClient(this.walletKeystore, password);
    await this.startScanning();
  }

  detachWallet() {
    this._storageService.remove('tn');
    this._storageService.remove('attachedDAYAccount');
    this.attachedDAYAccount = '';
    this.walletKeystore = '';
    this.stopScanning();
    this.eacWorker = null;
    showNotification('Your wallet has been detached.', 'success');
  }

  passwordMatchesKeystore(password) {
    try {
      ethereumJsWallet.fromV3(this.decrypt(this.walletKeystore), password, true);
      showNotification('Success.', 'success');
      return true;
    } catch (e) {
      if (e.message === 'Key derivation failed - possibly wrong passphrase') {
        showNotification('Please enter a valid password.');
      } else {
        showNotification(e);
      }
      return false;
    }
  }
}
