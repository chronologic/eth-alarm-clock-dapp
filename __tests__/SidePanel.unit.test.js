import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'mobx-react';
import Web3Service from '../app/services/web3';
import SidePanel from '../app/components/SidePanel/index';
import { Router } from 'react-router-dom';
import createBrowserHistory from 'history/createBrowserHistory';
import { RouterStore, syncHistoryWithStore } from 'mobx-react-router';
import TimeNodeStore from '../app/stores/TimeNodeStore';
import LocalStorageMock from '../__mocks__/LocalStorageMock';
import LocalStorageService from '../app/services/storage';
import EacStore from '../app/stores/EacStore';
import { eacService } from '../__mocks__/EacServiceMock';

const browserHistory = createBrowserHistory();
const routingStore = new RouterStore();

const history = syncHistoryWithStore(browserHistory, routingStore);

describe('SidePanel', () => {
  it('is correctly rendered', () => {
    const CURRENT_BLOCK = 5361200;

    const web3Service = new Web3Service({
      web3: {
        eth: {
          getBlockNumber(callback) {
            callback(null, CURRENT_BLOCK);
          }
        }
      }
    });

    const keenStore = {};
    const transactionStatistics = {};
    const featuresService = {};

    const localStorageMock = new LocalStorageMock();
    const storageService = new LocalStorageService(localStorageMock);
    const timeNodeStore = new TimeNodeStore({}, {}, {}, storageService);

    const eacStore = new EacStore(eacService, featuresService, web3Service);

    const injectables = {
      eacStore,
      keenStore,
      timeNodeStore,
      transactionStatistics,
      web3Service
    };

    const props = {
      location: {
        pathname: 'test',
        search: ''
      }
    };

    let mockedRender = renderer.create(
      <Provider {...injectables}>
        <Router history={history}>
          <SidePanel {...props} />
        </Router>
      </Provider>
    );

    let tree = mockedRender.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
