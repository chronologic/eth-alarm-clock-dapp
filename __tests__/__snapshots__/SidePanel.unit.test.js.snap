// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidePanel is correctly rendered 1`] = `
<nav
  className="page-sidebar"
  data-pages="sidebar"
>
  <div
    className="sidebar-header"
  >
    <img
      alt="logo"
      className="brand"
      data-src="img/logo-white.png"
      height="36"
      src="img/logo-white.png"
    />
    <div
      className="sidebar-header-controls"
    >
      <button
        className="btn btn-link d-lg-inline-block d-xlg-inline-block d-md-inline-block d-sm-none d-none"
        data-toggle-pin="sidebar"
        type="button"
      >
        <i
          className="fa fa-lock"
        />
      </button>
    </div>
  </div>
  <div
    className="sidebar-menu"
  >
    <ul
      className="menu-items"
    >
      <li
        className="m-t-30 "
      >
        <a
          aria-current={null}
          href="/"
          onClick={[Function]}
        >
          <span
            className="title "
          >
            Schedule
          </span>
          <span
            className="icon-thumbnail "
          >
            <i
              className="pg-calender"
            />
          </span>
        </a>
      </li>
      <li>
        <a
          href="#"
          onClick={[Function]}
        >
          <span
            className="title sub-title "
          >
            Transactions
          </span>
          <span
            className="icon-thumbnail "
          >
            <i
              className="pg-charts"
            />
          </span>
        </a>
        <ul
          className="sub-menu"
        >
          <li>
            <a
              aria-current={null}
              href="/transactions/scheduled"
              onClick={[Function]}
            >
              <span
                className="title"
              >
                Scheduled
              </span>
              <span
                className="icon-thumbnail"
              >
                <i
                  className="pg-plus_circle"
                />
              </span>
            </a>
          </li>
          <li>
            <a
              aria-current={null}
              href="/transactions/completed"
              onClick={[Function]}
            >
              <span
                className="title"
              >
                Completed
              </span>
              <span
                className="icon-thumbnail"
              >
                <i
                  className="fa fa-check"
                />
              </span>
            </a>
          </li>
        </ul>
      </li>
      <li>
        <a
          aria-current={null}
          href="/timenode"
          onClick={[Function]}
        >
          <span
            className="title "
          >
            TimeNode
          </span>
          <span
            className="icon-thumbnail "
          >
            <i
              className="fa fa-sitemap"
            />
          </span>
        </a>
      </li>
      <li>
        <a
          href="https://alpha.chronologic.network/debt/"
          rel="noopener noreferrer"
          target="_blank"
        >
          <span
            className="title"
          >
            Debt Smart Contract
          </span>
          <span
            className="icon-thumbnail"
          >
            <i
              className="fab fa-ethereum"
            />
          </span>
        </a>
      </li>
      <li>
        <a
          href="https://alpha.chronologic.network/chronos/"
          rel="noopener noreferrer"
          target="_blank"
        >
          <span
            className="title"
          >
            Day Token Contract
          </span>
          <span
            className="icon-thumbnail"
          >
            <i
              className="far fa-clock"
            />
          </span>
        </a>
      </li>
      <li>
        <a
          aria-current={null}
          href="/faucet"
          onClick={[Function]}
        >
          <span
            className="title "
          >
            Faucet
          </span>
          <span
            className="icon-thumbnail "
          >
            <i
              className="fas fa-tint"
            />
          </span>
        </a>
      </li>
      <li>
        <a
          href="https://blog.chronologic.network/chronos-platform/home"
          rel="noopener noreferrer"
          target="_blank"
        >
          <span
            className="title"
          >
            Help
          </span>
          <span
            className="icon-thumbnail"
          >
            <i
              className="far fa-question-circle"
            />
          </span>
        </a>
      </li>
      <hr
        className="d-md-block d-lg-none mx-4"
        id="sidebar-separator"
      />
      <li
        className="d-md-block d-lg-none"
      >
        <div
          className="container py-2"
        >
          <div
            className="row p-l-20 p-r-15"
          >
            <div
              className="col-8 px-0"
            >
              <span
                className="active-timenodes"
              >
                Active TimeNodes
              </span>
            </div>
            <div
              className="col-4 px-0 text-right"
            >
              <span
                className="timenode-count col-6"
              />
            </div>
          </div>
        </div>
      </li>
      <li
        className="d-md-block d-lg-none"
      >
        <div
          className="container py-2"
        >
          <div
            className="row p-l-20 p-r-15"
          >
            <div
              className="col-8 px-0"
            >
              <span
                className="active-timenodes"
              >
                Current Block
              </span>
            </div>
            <div
              className="col-4 px-0 text-right"
            >
              <span
                className="timenode-count col-6"
              />
            </div>
          </div>
        </div>
      </li>
    </ul>
    <div
      className="clearfix"
    />
  </div>
</nav>
`;
